### ISSUE TITLE
reduce api

### ISSUE THREAD

**Issue by sudopluto** [ejc-dev/c-array-impl#7]
"Arrays should grow when values are pushed back, and otherwise support reads and writes." - from assignment description

I think the api can be reduced to just support
push_back()
get_at()
set_at()
clear()

**Comment by Coderrine** [comment]
The only part of the API that definitely should be deleted is getContents (which is not even CwC). Getting rid of the other functions would restrict the usability of Arrays by a lot:

-Unable to remove a single element from an Array
-Unable to insert/add an element into an Array
-Unable to combine Arrays
-Unable to search an Array for a given Object
-Unable to get the size of an Array

**Comment by sudopluto** [comment]
I should have been more specific. How about:

Array() {}
~~size_t hash_;~~
void push_back(String* e) {} // Appends e to end
~~void add(size_t i, String* e) {} // Inserts e at i~~
~~void add_all(size_t i, StrList* c) {} // Inserts all of elements in c into this list at i~~
void clear() {} // Removes all of elements from this list
bool equals(Object* o) {} // Compares o with this list for equality.
String*  get(size_t index){} // Returns the element at index
size_t hash() {}  // Returns the hash code value for this list.
~~size_t index_of(Object* o) {} // Returns the index of the first occurrence of o, or >size() if not there~~
~~String* remove(size_t i) {} //Removes the element at i~~
String* set(size_t i, String* e)  {}// Replaces the element at i with e
size_t size() {}
~~std::vector<String*> getContents(){}~~
~~virtual size_t hash_me() {};~~

Right now this array looks more like an arraylist from java or a list from python.

**Comment by Coderrine** [comment]
The example you provided does not address my concerns:
If we need to remove something we added to an array, we can't.
If we need to insert something into an array not at the end (without replacing), we can't
If we need to combine Arrays, we can't.
If we need to search an Array and get the index of an Object, we can't.

Why would it be a problem if it resembles arraylist/list? The APIs and implementations should be as useful and adaptable as possible, since we are not being told how we will be using them. Of course, the creators of this API are free to share their thoughts as well.

**Comment by sudopluto** [comment]
Because array != arraylist. The API should follow conventions for data structures:

java: https://docs.oracle.com/javase/7/docs/api/java/util/Arrays.html
c++: http://www.cplusplus.com/reference/vector/vector/

If we need an arraylist / a more "useful / adaptable" data structure in the future, then we can always make it then, when the actual need arises.

**Comment by Coderrine** [comment]
I agree. The API should follow the conventions for the data structure that we have been asked to design. 

In the second link you provided (http://www.cplusplus.com/reference/vector/vector/) They have functionality that you did not include. These are:

Insert (Add and Add All)
Pop (Remove)

Index_Of does not appear to have a corresponding functionality in typical Arrays, however I don't believe it would restrict anyone's implementation to have this functionality, rather I think it would be useful.

**Comment by sudopluto** [comment]
I agree with you point about pop(). How about pop_back() as a sort of inverse to push_back()?

Do not agree with insert(). It could be a helper function, along with index_of(), if you would like to support it independently.

API:

Array() {}
void push_back(String* e) {} // Appends e to end
String* pop_back() // Pops last element from end
void clear() {} // Removes all of elements from this list
bool equals(Object* o) {} // Compares o with this list for equality.
String*  get(size_t index){} // Returns the element at index
size_t hash() {}  // Returns the hash code value for this list.
String* set(size_t i, String* e)  {}// Replaces the element at i with e
size_t size() {}


Removed:

~~size_t hash_;~~
~~std::vector<String*> getContents(){}~~
~~virtual size_t hash_me() {};~~

Helper Functions / Not Required / Optional:

size_t index_of(Object* o) {} // Returns the index of the first occurrence of o, or >size() if not there
void add(size_t i, String* e) {} // Inserts e at i
void add_all(size_t i, StrList* c) {} // Inserts all of elements in c into this list at i
String* remove(size_t i) {} //Removes the element at i
